### PR TITLE
network: fix issue caused by too many waiting replies

### LINKED ISSUE #26743: systemd-networkd: configuring tunnels cause instability
### systemd version the issue has been seen with

251

### Used distribution

Ubuntu 22.10

### Linux kernel version used

5.19.0-35-generic

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

After a restart of the `systemd-networkd` I was expecting to see the interfaces reconfigured working properly.

### Unexpected behaviour you saw

After a restart of the `systemd-networkd` service, if tunnels are configured (gre + sit), interfaces like Ethernet lost its IPv4 and IPv6 routes.

### Steps to reproduce the problem

Using Ubuntu 22.10 with systemd 251 (251.4-1ubuntu7.1).
I am originally using `netplan` and was able to reproduce the issue as long as there were tunnels in the config. `netplan generate` created the files listed below.

1. Create the following files

`/etc/systemd/network/15-sit.netdev`
```
[NetDev]
Name=foo
Kind=sit

[Tunnel]
Independent=true
Local=a.b.c.d
Remote=w.x.y.z
```

`/etc/systemd/network/15-sit.network`
```[Match]
Name=foo

[Network]
LinkLocalAddressing=ipv6
Address=a:b:c:d::1/64
ConfigureWithoutCarrier=yes
```

`/etc/systemd/network/15-bar.netdev`
```[NetDev]
Name=bar
Kind=gre

[Tunnel]
Independent=true
Local=a.b.c.d
Remote=w.x.y.z
```

`/etc/systemd/network/15-bar.network`
```[Match]
Name=bar

[Network]
LinkLocalAddressing=ipv6
Address=a:b:c:d::1/64
ConfigureWithoutCarrier=yes
```

2. Configure an Ethernet interface with IPv4, IPv6 (no DHCP)
3. Configure additional routes into `lo` interface
4. Restart the system
5. Everything should be configured and running normally
6. Run `sudo netplan --debug apply` or `systemctl restart systemd-networkd` at least once, wait for command to complete.

The `netplan` command will display something like:
```** (process:2271): DEBUG: 06:06:00.608: lo: adding new route
** (process:2271): DEBUG: 06:06:00.608: We have some netdefs, pass them through a final round of validation
** (process:2271): DEBUG: 06:06:00.608: eno1: setting default backend to 1
** (process:2271): DEBUG: 06:06:00.608: Configuration is valid
** (process:2271): DEBUG: 06:06:00.608: lo: setting default backend to 1
** (process:2271): DEBUG: 06:06:00.608: Configuration is valid
DEBUG:Merged config:
b''
Failed to reconfigure network interface eno1: Numerical result out of range
WARNING:Falling back to a hard restart of systemd-networkd.service
```
and sometimes will hang until cancelled to
```
subprocess.CalledProcessError: Command '['networkctl', 'reconfigure', '3', '1', '2', '5', '10', '9', '6', '8', '4', '7']' returned non-zero exit status 1.
```

And the logs of `systemd-networkd` will fill up with
```
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
```

There is a high chance static addresses and routes disappeared on the Ethernet interface. The service will keep failing and the presence or not of configuration files do not affect it.
Quickest remediation is to restart the system.

When trying without the 4 files and after a restart running
```bash
sudo ip tunnel add foo mode sit remote w.x.y.z local a.b.c.d
sudo ip -6 addr add a:b:c:d::1/64 dev foo
sudo ip link set foo up
sudo ip tunnel add bar mode gre remote w.x.y.z local a.b.c.d
sudo ip -6 addr add a:b:c:d::1/64 dev bar
sudo ip link set bar up
```

The service restarts and apply do not affect the interfaces.

Thank you for your help!

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Prompted by #26743.
Hopefully fixes #26743.

### CODE REVIEW

File: src/libsystemd/sd-netlink/sd-netlink.c [hunk 1]
```diff
@@ -458,6 +459,25 @@ static int timeout_compare(const void *a, const void *b) {
         return CMP(x->timeout, y->timeout);
 }
 
+int netlink_set_reply_callbacks_max(sd_netlink *nl, size_t n) {
+        assert(nl);
+
+        /* The default value is enough large. Refuse larger values. */
```
Comment (daandemeyer) [inline src/libsystemd/sd-netlink/sd-netlink.c]: "large enough"

**Review by DaanDeMeyer — APPROVED** [review]

File: src/libsystemd/sd-netlink/sd-netlink.c [hunk 2]
```diff
@@ -458,6 +459,25 @@ static int timeout_compare(const void *a, const void *b) {
...
+
+        nl->reply_callbacks_max = n;
+        return 0;
+}
+
+bool netlink_reply_callbacks_is_almost_full(sd_netlink *nl) {
+        assert(nl);
+
+        return hashmap_size(nl->reply_callbacks) >= nl->reply_callbacks_max - nl->reply_callbacks_max / 10;
+}
```
Comment (keszybz) [inline src/libsystemd/sd-netlink/sd-netlink.c]: This seems very ad-hoc. We don't export sd-netlink (which is probably something to revisit), but those functions are designed to be exported at some point. I don't think we'd want to export a heuristic like that. For example, for `n < 10` it only returns true if the array is full.

File: src/libsystemd/sd-netlink/sd-netlink.c [hunk 3]
```diff
@@ -35,6 +35,7 @@ static int netlink_new(sd_netlink **ret) {
                 .sockaddr.nl.nl_family = AF_NETLINK,
                 .original_pid = getpid_cached(),
                 .protocol = -1,
+                .reply_callbacks_max = DEFAULT_REPLY_CALLBACKS_MAX,
```
Comment (keszybz) [inline src/libsystemd/sd-netlink/sd-netlink.c]: Do I understand this correctly that the kernel is not happy with too many replies sent at once?
But if yes, then why set this to something that is so large? If all callers would want to set it to something much smaller, why not set it so?

**Review by keszybz — COMMENTED** [review]
Overall, I'm confused. Can you explain what sets the limit and why?

File: src/libsystemd/sd-netlink/sd-netlink.c (hunk 3 above)
Comment (yuwata) [reply src/libsystemd/sd-netlink/sd-netlink.c]: Now this is dropped.

File: src/libsystemd/sd-netlink/sd-netlink.c (hunk 2 above)
Comment (yuwata) [reply src/libsystemd/sd-netlink/sd-netlink.c]: This is dropped.

**Comment by yuwata** [comment]
@keszybz 
> Overall, I'm confused. Can you explain what sets the limit and why?

From https://github.com/systemd/systemd/issues/26743#issuecomment-1833359319,
> The difference is we have ~921044 routes and networkd is throwing same message for each route.

In such case, networkd tries to remove almost all routes when started, especially in a single event, then `sd_netlink.reply_callbacks` becomes full (921044 > UINT16_MAX).

**Comment by yuwata** [comment]
PTAL.

**Comment by yuwata** [comment]
> @yuwata Do you think this should go into v255?

I do not think so. But, it is better to backport the commits later.

File: src/libsystemd/sd-netlink/sd-netlink.c [hunk 4]
```diff
@@ -458,6 +458,12 @@ static int timeout_compare(const void *a, const void *b) {
         return CMP(x->timeout, y->timeout);
 }
 
+size_t netlink_get_num_reply_callbacks(sd_netlink *nl) {
```
Comment (keszybz) [inline src/libsystemd/sd-netlink/sd-netlink.c]: As Lennart would say, avoid needless abbreviations.

`netlink_get_reply_callback_count`?

File: src/shared/firewall-util.c [hunk 5]
```diff
@@ -90,6 +91,13 @@ FirewallContext *fw_ctx_free(FirewallContext *ctx) {
         return mfree(ctx);
 }
 
+size_t fw_ctx_get_num_reply_callbacks(FirewallContext *ctx) {
```
Comment (keszybz) [inline src/shared/firewall-util.c]: `fw_ctx_get_reply_callback_count`?

**Comment by yuwata** [comment]
> Thanks! This is much more obvious compared to the previous version. If we can get a confirmation that this solves the issue, I think we could merge this for v255.

I do not think we can get any confirmation from the reporters of the issue, as I guess the issue is caused by mis-configuration, and at least one case is already fixed by `ManageForeignRoutes=no`.
Dropping "waiting-for-reporter-feedback" tag.

**Comment by YHNdnzj** [comment]
(Some of the commit messages need updating BTW.)

**Pull request merged**